debridge_node/src/modules/chain/scanning/services/SolanaReaderService.ts

```diff
@@ -1,161 +1,127 @@
-          return submission;
-        } catch (e) {
-          this.logger.error(`Can't generate submission ${event.submissionId} from solana event`);
-          this.logger.error(e);
-          throw e;
+    for await (const response of this.#duplex.responses) {
+      if (!(response.sendEventMessage !== undefined && response.sendEventMessage.oneofKind !== undefined)) {
+        break;
+      }
+      switch (response.sendEventMessage.oneofKind) {
+        case undefined: {
+          continue;
         }
-      });
-
-      //sort in asc, we need it for correct updating last tracked value and nonce validation
-      submissions.sort((a, b) => a.nonce - b.nonce);
-      this.logger.verbose(`Events ${submissions.length} are prepared for db storing`);
+        case 'heartbeat': {
+          // @ts-ignore
+          this.#logger.verbose(`Heartbeat: ${JSON.stringify(response.sendEventMessage.heartbeat)}`);
 
-      if (submissions.length > 0) {
-        const result = await this.chainProcessingService.process(submissions, chainId, submissions.at(-1).txHash);
-        if (result !== ProcessNewTransferResultStatusEnum.SUCCESS) {
+          this.heartbeat();
           break;
         }
-      } else {
-        this.logger.log(`There are no success transaction`);
-        const lastTransactionInBatch = transactions.at(-1);
-        if (lastTransactionInBatch) {
-          await this.supportedChainRepository.update(chainId, {
-            latestSolanaTransaction: lastTransactionInBatch.transactionHash,
-            latestBlock: lastTransactionInBatch.slotNumber,
-            lastTxTimestamp: lastTransactionInBatch.transactionTimestamp.toString(),
-            lastTransactionSlotNumber: lastTransactionInBatch.slotNumber,
-          });
-          this.logger.log(
-            `updateSupportedChainBlock; transactionHash: ${lastTransactionInBatch.transactionHash}; slotNumber: ${lastTransactionInBatch.slotNumber};`,
-          );
+        case 'event': {
+          this.heartbeat();
+          // @ts-ignore
+          const event = response.sendEventMessage.event;
+          const nonce = Number(U256Converter.toBigInt(event.submission?.nonce).toString());
+          this.#logger.log(`new event is received nonce: ${nonce}`);
+
+          const submission = this.transformService.generateSubmissionFromSolanaSendEvent(event);
+          this.#submissionsFromSync.push(submission);
+          break;
         }
       }
-      this.logger.verbose(`Waiting ${this.SOLANA_API_WAIT_BATCH_INTERVAL}`); //need for sync events from reader
-      await setTimeout(this.SOLANA_API_WAIT_BATCH_INTERVAL);
     }
   }
+
+  private heartbeat() {
+    clearTimeout(this.#heartbeatIntervalInstance);
+    const timeout = this.configService.get<number>('DEBRIDGE_EVENTS_CONSISTENCY_CHECK_TIMEOUT_SECS') * 10 * 1000;
+    this.#heartbeatIntervalInstance = setTimeout(async () => {
+      this.createSubscription();
```

Comment:
You don't need to re-create subscription on every heartbeat, grpc already have `heartbeat` events.
Use setInterval with binded value (e.g. isAlive), update value on every event (heartbeat or data) `isAlive = true`, throw error if interval ticks but isAlive = false

```ts
checkHeartbeat() {
if (!this.isAlive) {
  throw new Error();
}
this.isAlive = false; // will be set to true when event will be received
}

setInterval(this.checkHeartbeat.bind(this), timeout)
```